## Unit 2:  Homework Option III -- Working With Large Files

This homework provides students with the option to get practice opening, sampling, and cleaning data when it has file sizes that might stretch your available RAM, or even exceed it.  

**What you'll turn in:**

You'll make a pull request with a folder (titled as your name) that contains the following:

 - for section I, a file called `titanic`, stored in a binary `feather` or `parquet` file format, that reduces the `titanic` dataset to its smallest possible memory amount without tampering any of its values.
 - for section II, simply turn in this notebook with the questions answered.
 - for section III, a script called `chunking.py` that, when run, will connect to an s3 bucket, and stream a file through memory, clean it, and output it to a `.csv` file, even though the entire file was never entirely loaded into RAM.

### Section I: Downcasting the Titanic Dataset

**What You'll Learn:** The basics of managing memory within files, and how to use advanced file formats such as the `feather` or `parquet`, to make it easier to maintain persistent data types when a file isn't loaded.

**What You'll Turn In:** A file called `titanic`, which contains the memory-reduced form of the file.

#### Downcasting

`Downcasting` is the task of reducing the memory footprint of different columns in your dataset so they take up less RAM when you load them in.  

Most software used for handling data makes use of your available RAM to process its tasks.  If the size of your file neatly fits into the available RAM that your computer has then this is fine.  If it's significantly larger (no laptop is going to have 1TB of RAM for example), then you won't be able to load in the file and work with it.

Pandas works this way, and therefore the amount of working RAM you have available to use is going to function as a limit for what file sizes you can work with.  

##### A Quick Intro to Data Types in Pandas & Numpy

Numbers come in different flavors in pandas and numpy.  At the simplest level you have integers (whole numbers) and floating point numbers (numbers with decimals).  

However, numbers use different sized containers to store their values.  They are as follows:

 - **64 bit:** Can store values as large as 2<sup>64</sup>, which is 18446744073709551616
 - **32 bit:** Can store values as large as 2<sup>32</sup>, which is 4294967296
 - **16 bit:** Good for values up to 2<sup>16</sup>, 65536
 - **8 bit:**  2<sup>8</sup>, or up to 264
 
**Integers** typically have the above range, while **floats** typically can only go down to 32 bits.

You can see the whole range of numeric data types here:  https://docs.scipy.org/doc/numpy/user/basics.types.html

The important detail here, is that if a number is encoded as being a 64 bit number, it will *always* use the same amount of memory to store it, even if the value itself is much smaller.  

So, if you have a column of 0's and 1's, an 8 bit encoding will work perfectly fine (since the values are less than 2<sup>8</sup>), and a 64 bit encoding will take up 8x as much memory as it needs to.

An important detail about how Pandas works is that *all numbers are automatically encoded as 64 bit numbers*.  This is good for making sure values aren't tampered with, but bad for optimizing memory with large files.

**Methods Used For Managing Memory In Pandas:**

 - `df.memory_usage()`, returns the memory usage, in bytes, of whatever is selected.
 - `pd.Series.astype()`, allows you to change the data type of 1 variable to another.
 - `df.info()`, returns the data type and memory usage of every column selected
 - `df.dtypes()/pd.Series.dtype()`, returns the data type of everything selected
 
Let's take a look at how these items work.  Run the following cells to get a quick demonstration of what they do.

In [5]:
!pwd

Start setting keys..
Keys set.
/Users/PRSmb/OneDrive/General-Assembly/my-1019-repo/Homework/Unit2


In [1]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
import plotly.express as px
import plotly.io as pio
pio.templates

Templates configuration
-----------------------
    Default template: 'plotly'
    Available templates:
        ['ggplot2', 'seaborn', 'simple_white', 'plotly',
         'plotly_white', 'plotly_dark', 'presentation', 'xgridoff',
         'ygridoff', 'gridon', 'none']

In [77]:
# load in the titanic dataset here -- use a different url if need be to
# load it in
#import pandas as pd
#import numpy as np
df = pd.read_csv('./data/titanic.csv')

In [78]:
# check the info of your dataset -- notice the 64 bit numbers
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [80]:
# We can also check the memory usage of each column
df.memory_usage()

Index           128
PassengerId    7128
Survived       7128
Pclass         7128
Name           7128
Sex            7128
Age            7128
SibSp          7128
Parch          7128
Ticket         7128
Fare           7128
Cabin          7128
Embarked       7128
dtype: int64

In [81]:
df.memory_usage().sum()

85664

In [7]:
# now let's see what happens when we adjust a columns data type
df['Survived'].astype(np.int8).memory_usage()

1019

As you can see, an 8 bit number takes up about 1/8 as much memory as a 64 bit number.  Permanently changing something's data type is simple.

In [8]:
# this will change the data type of a column to something else
df['Survived'] = df['Survived'].astype(np.int8)

In [9]:
# and now we can see its memory footprint is permanently smaller
df.memory_usage()

Index           128
PassengerId    7128
Survived        891
Pclass         7128
Name           7128
Sex            7128
Age            7128
SibSp          7128
Parch          7128
Ticket         7128
Fare           7128
Cabin          7128
Embarked       7128
dtype: int64

Notice however, that if you make a columns data type *smaller* than what it is, the original values will be tampered.  For example, the Passenger ID column has values as large as 891.....which is more than 2<sup>8</sup>.  Notice what happens when you make the change:

In [10]:
# the values at the end of the series should be 889, 890, 891, etc
df['PassengerId'].astype(np.int8)

0        1
1        2
2        3
3        4
4        5
      ... 
886    119
887    120
888    121
889    122
890    123
Name: PassengerId, Length: 891, dtype: int8

So clearly, getting things 'just right' is important.  Notice also the difference between **signed** and **unsigned** data types.  If it's signed, that means they can accept negative values.  

So, a datatype of `np.uint8` can accept ranges of 0 - 255, whereas `np.int8` accepts values from -128 to 127.

##### Categorical Data in Pandas and Numpy

Text based data in Pandas and Numpy has two different varieties:

 - **np.object**: this is the default numpy way of treating and handling data.  Pretty close to a python string, and is used to store data that doesn't have any other characteristic (integer, float, bool, etc)
 - **category**: this is a special data type built specifically for Pandas, to handle text data that has a small number of repeating values.  Like the `sex` column in our dataset.  When appropriately used, it can drastically reduce the memory footprint of text based data.  You can read more about them here:  https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html
 
See below for a demonstration.

In [11]:
# the Sex column currently uses the same memory as a 64 bit number
df['Sex'].memory_usage()

7256

In [12]:
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [13]:
# if you change it to a category, it uses about the same amount
# as an 8 bit number
df['Sex'].astype('category').memory_usage()

1115

Please note that using categories only has the desired effect when there are repeating values.  Make sure to check the memory footprint before making the change!

#### Section I Task:  Clean Up the Titanic Dataset

For the first part of your homework assignment, your task is to reduce the memory footprint of the titanic dataset as much as possible, while not tampering any values, and export it to a binary file format called `feather`(preferably), or `parquet` which will maintain information about all of its data types when it's loaded back in. 

You can read more about the feather file format here:  https://blog.rstudio.com/2016/03/29/feather/.

The file `titanic` should be turned in inside a folder with your name on it, and it will be inspected to make sure it was downcasted in the most appropriate manner.

You can use the `to_feather()`, and `read_feather()` methods to load and export your files.  Or the `to_parquet()` and `read_parquet()` methods, respectively.

**Note:** It is best to use feather files if you can.  They are the only file format that can maintain persistent information about the `category` data type after it's been saved.

However, you will likely need some additional libraries to get it working.  Often pandas does not work well with `feather` files right out of the box.  The typical library used to work with feather files is, appropriately called......feather.  To install the `feather` library, simply go to Anaconda Prompt/Terminal and type in `conda install feather-format`.

The library's homepage can be found here:  https://pypi.org/project/feather-format/

This library isn't always well supported, and there's a possibility that you might have an exceedingly difficult time getting it to work.  The purpose of this homework is not to have you go down rabbit holes getting an obscure library to work.  If after 30 minutes - 1 hour you still don't have the file installed, feel free to use the `parquet` file format, which can be a little easier to work with.

You can find more about it here:  https://arrow.apache.org/docs/python/parquet.html

### Section 1 (1 of 2) - Reduce file size

In [14]:
# Starting point
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int8   
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(4), int8(1), object(5)
memory usage: 77.6+ KB


In [15]:
df.memory_usage()

Index           128
PassengerId    7128
Survived        891
Pclass         7128
Name           7128
Sex            7128
Age            7128
SibSp          7128
Parch          7128
Ticket         7128
Fare           7128
Cabin          7128
Embarked       7128
dtype: int64

In [ ]:
# check min and max value of each numeric column

In [16]:
df.min()

PassengerId                      1
Survived                         0
Pclass                           1
Name           Abbing, Mr. Anthony
Sex                         female
Age                           0.42
SibSp                            0
Parch                            0
Ticket                      110152
Fare                             0
dtype: object

In [17]:
df.max()

PassengerId                            891
Survived                                 1
Pclass                                   3
Name           van Melkebeke, Mr. Philemon
Sex                                   male
Age                                     80
SibSp                                    8
Parch                                    6
Ticket                           WE/P 5735
Fare                               512.329
dtype: object

**Based on workings:**
* PassengerId 1-891 => np.int16

* Survivied 0-1 => np.bool_

* Pclass 1-3 => Test category or int8; use smaller

* Name => This is text so don't mess with it as yet

* Sex | check all values likely 'male','female' => Male or Female, test category, or boolean (since could deefine as male_flag without losing info, but potentially losing .

* SibSp 0 - 8 => Likely repeating values, => test with category but probably int8

* Age 0 - 80 => Likely repeating values, test with category but probably float16

* Parch 0 - 6 => Likely repeating values, test with category but probably int8

* Ticket => This is text so don't mess with it as yet

* Fare 0 to 512.329 => Test with np.float16 (half precision float)

* Cabin => Lots of text and nulls so no change

* Embarked => Limited set of locations, so likely category

In [24]:
test_df = df.copy()

In [30]:
a = df['PassengerId'].memory_usage()
b = df['PassengerId'].astype(np.int16).memory_usage()
print(f'Initial: {a}\nReduced: {b}')

Initial: 7256
Reduced: 1910


In [31]:
a = df['Survived'].memory_usage()
b = df['Survived'].astype(np.bool_).memory_usage()
print(f'Initial: {a}\nReduced: {b}')

Initial: 1019
Reduced: 1019


In [33]:
a = df['Pclass'].memory_usage()
b = df['Pclass'].astype(np.int8).memory_usage()
c = df['Pclass'].astype('category').memory_usage()
print(f'Initial: {a}\nReduced: {b}\nReduced - Category: {c}')

Initial: 7256
Reduced: 1019
Reduced - Category: 1123


In [34]:
a = df['Sex'].memory_usage()
b = df['Sex'].astype('category').memory_usage()
print(f'Initial: {a}\nReduced: {b}')

Initial: 7256
Reduced: 1115


In [44]:
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [40]:
test_df['male_flag'] = (test_df['Sex']=='male')

In [41]:
test_df['male_flag'].memory_usage() 

1019

Forcing a boolean flag of male / female would reduce memory usage beyond category by a further 96 bytes, but would force a change of data from the original data set and potentially reduce clarity. So won't do this.

In [49]:
a = df['SibSp'].memory_usage()
b = df['SibSp'].astype(np.int8).memory_usage()
c = df['SibSp'].astype('category').memory_usage()
print(f'Initial: {a}\nReduced: {b}\nReduced - Category: {c}')

Initial: 7256
Reduced: 1019
Reduced - Category: 1395


In [51]:
a = df['Age'].memory_usage()
b = df['Age'].astype(np.float16).memory_usage()
print(f'Initial: {a}\nReduced: {b}\n')

Initial: 7256
Reduced: 1910



In [54]:
a = df['Age'].memory_usage()
b = df['Age'].astype(np.float16).memory_usage()
print(f'Initial: {a}\nReduced: {b}\n')

Initial: 7256
Reduced: 1910



In [55]:
# Age has nulls, so when try to use integers it fails
df['Age'].isnull().sum()

177

In [59]:
df['Age'].astype(np.int8).memory_usage() # This produces errors

ValueError: Cannot convert non-finite values (NA or inf) to integer

In [60]:
a = df['Parch'].memory_usage()
b = df['Parch'].astype(np.int8).memory_usage()
c = df['Parch'].astype('category').memory_usage()
print(f'Initial: {a}\nReduced: {b}\nReduced - Category: {c}')

Initial: 7256
Reduced: 1019
Reduced - Category: 1395


In [66]:
a = df['Fare'].memory_usage()
b = df['Fare'].astype(np.float16).memory_usage()

print(f'Initial: {a}\nReduced: {b}')

Initial: 7256
Reduced: 1910


In [67]:
a = df['Embarked'].memory_usage()
b = df['Embarked'].astype('category').memory_usage()

print(f'Initial: {a}\nReduced: {b}')

Initial: 7256
Reduced: 1123


In [69]:
# So best layout with new data types:

df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [82]:
df['PassengerId'] = df['PassengerId'].astype(np.int16)
df['Survived'] = df['Survived'].astype(np.bool_)
df['Pclass'] = df['Pclass'].astype(np.int8)
df['Name'] = df['Name'].astype(str)
df['Sex'] = df['Sex'].astype('category')
df['Age'] = df['Age'].astype(np.float16)
df['SibSp'] = df['SibSp'].astype(np.int8)
df['Parch'] = df['Parch'].astype(np.int8)
df['Ticket'] = df['Ticket'].astype(str)
df['Fare'] = df['Fare'].astype(np.float16)
df['Cabin'] = df['Cabin'].astype(str)
df['Embarked'] = df['Embarked'].astype('category')

In [83]:
df.memory_usage()

Index           128
PassengerId    1782
Survived        891
Pclass          891
Name           7128
Sex             987
Age            1782
SibSp           891
Parch           891
Ticket         7128
Fare           1782
Cabin          7128
Embarked        995
dtype: int64

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    int16   
 1   Survived     891 non-null    bool    
 2   Pclass       891 non-null    int8    
 3   Name         891 non-null    object  
 4   Sex          891 non-null    category
 5   Age          714 non-null    float16 
 6   SibSp        891 non-null    int8    
 7   Parch        891 non-null    int8    
 8   Ticket       891 non-null    object  
 9   Fare         891 non-null    float16 
 10  Cabin        891 non-null    object  
 11  Embarked     889 non-null    category
dtypes: bool(1), category(2), float16(2), int16(1), int8(3), object(3)
memory usage: 31.6+ KB


In [85]:
df.memory_usage().sum()

32404

So - in terms of adjusting data types, we were able to reduce memory from **85,664** to **32,404** within Pandas. This is a ~62% reduction

### Section 1 (2 of 2) - Create Feather File

In [87]:
import feather

In [98]:
# Uses LZ4 by default
feather.write_dataframe(df, './feather/titantic-f-default')

# # Use LZ4 explicitly
feather.write_dataframe(df, './feather/titantic-f-force-lz4', compression='lz4')

# # Use ZSTD
feather.write_dataframe(df, './feather/titantic-f-zstd', compression='zstd')

# # Do not compress
feather.write_dataframe(df, './feather/titantic-f-uncompressed', compression='uncompressed')

In [115]:
! pwd

Start setting keys..
Keys set.
/Users/PRSmb/OneDrive/General-Assembly/my-1019-repo/Homework/Unit2


In [118]:
! ls -lh ./feather 

Start setting keys..
Keys set.
total 376
-rw-r--r--@ 1 PRSmb  staff    45K Nov 22 13:01 titantic-f-default
-rw-r--r--@ 1 PRSmb  staff    45K Nov 22 13:01 titantic-f-force-lz4
-rw-r--r--@ 1 PRSmb  staff    58K Nov 22 13:01 titantic-f-uncompressed
-rw-r--r--@ 1 PRSmb  staff    31K Nov 22 13:01 titantic-f-zstd


So smallest version of the file size (after reducing the file size) seems to be a 31 KB feather file, using zstd compression.

### Section II: Working With Larger File Sizes

Section I was your warmup to get the hang of how to reduce the memory of your data and get it loaded into a more advanced file format that can be reused for other projects.

This section will extend what you just did, but add in two additional wrinkles:

 - The file will be much larger -- approximately 2 million rows
 - You'll find out what types of data types you should be working with.....*without* reading in the entire file in the first place.

Imagine this hypothetical scenario:  there's a 10 gb file that you need to work on, and loading it into pandas is returning a `MemoryError`.  You're almost certain you could reduce it to something much smaller......except you can't even read it in to figure out what to do next.

These types of chicken & egg problems are fairly common, and a popular antidote to them is to sample in a portion of your dataframe.  There are two primary ways to do this if you're reading in `.csv` files:

 - `nrows`   : tells you how many rows to read in from the original file
 - `skiprows`: tells you which rows to *skip* from the original file
 
For both of these you can just manually read in x number of rows relatively easily, but doing so right from the beginning or end has some problems.  Mainly, many datasets don't have consistent values from beginning to end.  

For example, if you have a dataset with 10 years of sales info, it's very possible values being recorded are very different at varying time segments.  Often new columns are added to datasets in the middle of their collection, and all values for all times before that are simply `null`, so just reading in the first 5000 or 10000 might not give you a consistent picture of what to expect.

For this reason, it's good to randomly sample in dataframes before you want to read them in entirely.

To do this, you can mix `lambda` functions (remember those?) with the `skiprows` argument, which can accept a function as values.

Here's the basic idea:

In [18]:
import numpy as np
import pandas as pd

In [7]:
import random

# random.random() will generate a random value between 0 and 1
# so this will return True 50% of the time
random.random() > .50

True

In [16]:
i = 0
for n in range(0,100000):
    if random.random()>.50: i = i + 1
        
print(f"{i/100000}")

0.49854


Using this same logic, we can pass this into a lambda function like so, to read in 30% of the titanic dataset.

In [24]:
# use read_csv with a lambda function
df = pd.read_csv('./data/titanic.csv', skiprows=lambda x: x > 0 and random.random() > 0.3)
# and notice the size of the dataframe that was read in
df.shape

(265, 12)

Here, inside the `lambda` function, `x` represents the index value of the row being read in.  `x > 0` is used assuming the first row are headers, and of course `random.random() > 0.3` will return `True` 70% of the time, hence the results that we get.

The idea is that you can read in a very small fraction of a very large file if you want to investigate its most important properties.  

Now, this begs the question.....how do you specify the data types you want a column to be before you read it in?  

There is a very useful argument in `read_csv` called `dtype` that accepts a dictionary, where you can list column labels as keys, and their corresponding data type as a value.

So for example, if we wanted to change the `Embarked` column and the `Survived` column to `category` and `np.int8`, we could do so in the following way.

In [26]:
# this dictionary contains the columns we want to change
dtypes = {
    # column label: new data type
    'Embarked': 'category',
    'Survived': np.int8
}

# and now we'll re-read in the .csv file, and use the dtypes dict
df = pd.read_csv('./data/titanic.csv', dtype=dtypes)

In [27]:
# and we can see that the data type of the columns are in fact changed
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    int64   
 1   Survived     891 non-null    int8    
 2   Pclass       891 non-null    int64   
 3   Name         891 non-null    object  
 4   Sex          891 non-null    object  
 5   Age          714 non-null    float64 
 6   SibSp        891 non-null    int64   
 7   Parch        891 non-null    int64   
 8   Ticket       891 non-null    object  
 9   Fare         891 non-null    float64 
 10  Cabin        204 non-null    object  
 11  Embarked     889 non-null    category
dtypes: category(1), float64(2), int64(4), int8(1), object(4)
memory usage: 71.6+ KB


#### Section II Task: Sample, Clean Up, And Read in the taxi.csv file

This portion of section II contains the task that you will be graded on.  You can simply answer the prompts inside this notebook and turn it in.  No additional files are necessary.  The file is located in an S3 bucket at this location:  `https://dat-data.s3.amazonaws.com/taxi.csv`

It records information about every taxi ride given by a particular company for approximately 1 year.

**Part I:** Randomly sample in 10% the taxi.csv file

In [ ]:
# your answer here

Manual examination of the file shows taxi.csv has: 
* 1,710,671 rows
* 8 columns

In [28]:
# use read_csv with a lambda function
df = pd.read_csv('./data/taxi.csv', skiprows=lambda x: x > 0 and random.random() > 0.1)
# and notice the size of the dataframe that was read in
df.shape

(171333, 8)

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 171333 entries, 0 to 171332
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   TRIP_ID       171333 non-null  int64  
 1   CALL_TYPE     171333 non-null  object 
 2   ORIGIN_CALL   36687 non-null   float64
 3   ORIGIN_STAND  80875 non-null   float64
 4   TAXI_ID       171333 non-null  int64  
 5   TIMESTAMP     171333 non-null  int64  
 6   DAY_TYPE      171333 non-null  object 
 7   MISSING_DATA  171333 non-null  bool   
dtypes: bool(1), float64(2), int64(3), object(2)
memory usage: 9.3+ MB


In [30]:
df.memory_usage()

Index               128
TRIP_ID         1370664
CALL_TYPE       1370664
ORIGIN_CALL     1370664
ORIGIN_STAND    1370664
TAXI_ID         1370664
TIMESTAMP       1370664
DAY_TYPE        1370664
MISSING_DATA     171333
dtype: int64

In [32]:
df.memory_usage().sum()

9766109

Given we pulled in 10% of the file, we can assume that without modifying things the full file will be around 90mb

**Part II:** Go ahead and do the appropriate exploratory data analysis to figure out the most appropriate data type for each column.

In [ ]:
# your answer here

In [33]:
df.head()

TRIP_ID CALL_TYPE  ORIGIN_CALL  ORIGIN_STAND   TAXI_ID  \
0  1372636965620000231         C          NaN           NaN  20000231   
1  1372638481620000403         B          NaN          28.0  20000403   
2  1372639875620000009         C          NaN           NaN  20000009   
3  1372642178620000060         C          NaN           NaN  20000060   
4  1372636896620000360         C          NaN           NaN  20000360   

    TIMESTAMP DAY_TYPE  MISSING_DATA  
0  1372636965        A         False  
1  1372638481        A         False  
2  1372639875        A         False  
3  1372642178        A         False  
4  1372636896        A         False

In [63]:
df.isnull().sum()

TRIP_ID              0
CALL_TYPE            0
ORIGIN_CALL     134646
ORIGIN_STAND     90458
TAXI_ID              0
TIMESTAMP            0
DAY_TYPE             0
MISSING_DATA         0
dtype: int64

**Initial observations** from looking at raw file, and header column

* TRIP_ID - A very long long int
* CALL_TYPE - Looks like some sort of category
* ORIGIN_CALL - Lots of empty values, need to do some value checks but try category first
* ORIGIN_STAND - Similar to origin_call
* TAXI_ID - Looks like a long int
* TIMESTAMP - Looks like a long int, not obviously a date time. Maybe it's already in UNIX logic?
* DAY_TYPE - Looks like lots of repeated values
* MISSING_DATA - Looks like it is boolean / lots of repeaated values

In [62]:
for c in df.columns:
    print(f"{c}\t\tmax: {df[c].max()}")

TRIP_ID		max: 1404172754620000518
CALL_TYPE		max: C
ORIGIN_CALL		max: 63882.0
ORIGIN_STAND		max: 63.0
TAXI_ID		max: 20000969
TIMESTAMP		max: 1404172754
DAY_TYPE		max: A
MISSING_DATA		max: False


In [71]:
a = df['TRIP_ID'].memory_usage()
b = df['TRIP_ID'].astype(np.int64).memory_usage()
print(f"Original:{a} New:{b}")

Original:1370792 New:1370792


In [80]:
a = df['CALL_TYPE'].memory_usage()
b = df['CALL_TYPE'].astype('category').memory_usage()
print(f"Original:{a} New:{b}")

Original:1370792 New:171565


In [74]:
a = df['ORIGIN_CALL'].memory_usage()
b = df['ORIGIN_CALL'].astype(np.float16).memory_usage()
print(f"Original:{a} New:{b}")

Original:1370792 New:342794


In [75]:
a = df['ORIGIN_STAND'].memory_usage()
b = df['ORIGIN_STAND'].astype(np.float16).memory_usage()
print(f"Original:{a} New:{b}")

Original:1370792 New:342794


In [78]:
a = df['TAXI_ID'].memory_usage()
b = df['TAXI_ID'].astype(np.float16).memory_usage()
print(f"Original:{a} New:{b}")

Original:1370792 New:342794


In [82]:
a = df['TIMESTAMP'].memory_usage()
b = df['TIMESTAMP'].astype(np.uint32).memory_usage()
print(f"Original:{a} New:{b}")

Original:1370792 New:685460


In [81]:
a = df['TIMESTAMP'].memory_usage()
b = df['TIMESTAMP'].astype(np.uint32).memory_usage()
print(f"Original:{a} New:{b}")

Original:1370792 New:685460


In [83]:
a = df['DAY_TYPE'].memory_usage()
b = df['DAY_TYPE'].astype('category').memory_usage()
print(f"Original:{a} New:{b}")

Original:1370792 New:171549


In [84]:
a = df['MISSING_DATA'].memory_usage()
b = df['MISSING_DATA'].astype(np.bool_).memory_usage()
print(f"Original:{a} New:{b}")

Original:171461 New:171461


**Part III**: Create a dictionary that contains the key/value pairs for each column that needs to be changed to a different data type, and then read in the file.

In [ ]:
# your answer here

In [89]:
np.finfo(np.float16)

finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)

In [90]:
np.finfo(np.float32)

finfo(resolution=1e-06, min=-3.4028235e+38, max=3.4028235e+38, dtype=float32)

In [97]:
# Note - Decided to use all columns to ensure didn't miss anything
dtypes_taxi = {
    # column label: new data type
    'TRIP_ID' : np.int64,
    'CALL_TYPE' : 'category',
    'ORIGIN_CALL' : np.float32, 
        # This causes an error set to np.uint16 due to base being a float
        # So use float32 since max value seen was 63882 in sample, could go above 66k
        # This is likely due to there being null / NaN values
    'ORIGIN_STAND' : np.float16,
        #This causes an error set to np.uint8 due to base being a float
        # So use float16 since max value seen was 2 digit 63.0 well under 65k
        # This is likely due to there being null / NaN values
    'TAXI_ID' : np.uint32,
    'TIMESTAMP' : np.uint32,
    'DAY_TYPE' : 'category',
    'MISSING_DATA' : np.bool_    ,
}

In [98]:
dtypes_taxi

{'TRIP_ID': numpy.int64,
 'CALL_TYPE': 'category',
 'ORIGIN_CALL': numpy.float32,
 'ORIGIN_STAND': numpy.float16,
 'TAXI_ID': numpy.uint32,
 'TIMESTAMP': numpy.uint32,
 'DAY_TYPE': 'category',
 'MISSING_DATA': numpy.bool_}

In [99]:
df_taxi = pd.read_csv('./data/taxi.csv', dtype=dtypes_taxi)

In [102]:
df_taxi.shape

(1710670, 8)

**Part IV:** Confirm that each column has the appropriate data type

In [ ]:
# your answer here

In [105]:
df_taxi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1710670 entries, 0 to 1710669
Data columns (total 8 columns):
 #   Column        Dtype   
---  ------        -----   
 0   TRIP_ID       int64   
 1   CALL_TYPE     category
 2   ORIGIN_CALL   float32 
 3   ORIGIN_STAND  float16 
 4   TAXI_ID       uint32  
 5   TIMESTAMP     uint32  
 6   DAY_TYPE      category
 7   MISSING_DATA  bool    
dtypes: bool(1), category(2), float16(1), float32(1), int64(1), uint32(2)
memory usage: 40.8 MB


In [106]:
df_taxi.memory_usage()

Index                128
TRIP_ID         13685360
CALL_TYPE        1710774
ORIGIN_CALL      6842680
ORIGIN_STAND     3421340
TAXI_ID          6842680
TIMESTAMP        6842680
DAY_TYPE         1710678
MISSING_DATA     1710670
dtype: int64

In [107]:
df_taxi.memory_usage().sum()

42766990

### Section III (Optional): File Streaming

File streaming is a way to get around memory limitations in pandas.  When you stream in a file, you spoon feed a portion of it into memory, and when you're finished, load in the next portion, and so on until there's nothing left. 

It's less convenient than regular file I/O, but it removes any sort of memory limit you might face when working with a file because it's never loaded at the same time. 

In this section of the homework assignment, you'll be tasked with performing basic cleaning operations on a file......without ever having to load it into memory.  This means what you accomplish in this section you could perform on a file of any size.

##### Basic Introduction to File Streams

When it's available, you can specify a file stream by using the `chunksize` argument, which specifies that you only read in so many lines at a time.

Notice how it works:

In [3]:
# when we read in the csv file, we'll set chunksize to 200
df = pd.read_csv('./data/titanic.csv', chunksize=250)

# notice that df is NOT a df.....it's a file stream
type(df)

pandas.io.parsers.TextFileReader

In [4]:
# and now if we want, we can 'chunk' in 250 rows at a time
df.get_chunk()

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
245          246         0       1   
246          247         0       3   
247          248         1       2   
248          249         1       1   
249          250         0       2   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
245                        Minahan, Dr. William Edward    male  44.0      2   
246              Lindahl, Miss. Agda Thorilda Viktoria  female  25.0      0   
247                    Hamalainen, Mrs. William (Anna)  female  24.0      0   
248                      Beckwith, Mr. Richard Leonard    male  37.0      1   
249                      Carter, Rev. Ernest Courtenay    male  54.0      1   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
245      0             19928  90.0000   C78        Q  
246      0            347071   7.7750   NaN        S  
247      2            250649  14.5000   NaN        S  
248      1             11751  52.5542   D35        S  
249      0            244252  26.0000   NaN        S  

[250 rows x 12 columns]

In [5]:
# and this would be the next 250 rows
df.get_chunk()

PassengerId  Survived  Pclass  \
250          251         0       3   
251          252         0       3   
252          253         0       1   
253          254         0       3   
254          255         0       3   
..           ...       ...     ...   
495          496         0       3   
496          497         1       1   
497          498         0       3   
498          499         0       1   
499          500         0       3   

                                                Name     Sex   Age  SibSp  \
250                           Reed, Mr. James George    male   NaN      0   
251       Strom, Mrs. Wilhelm (Elna Matilda Persson)  female  29.0      1   
252                        Stead, Mr. William Thomas    male  62.0      0   
253                         Lobb, Mr. William Arthur    male  30.0      1   
254         Rosblom, Mrs. Viktor (Helena Wilhelmina)  female  41.0      0   
..                                               ...     ...   ...    ...   
495                            Yousseff, Mr. Gerious    male   NaN      0   
496                   Eustis, Miss. Elizabeth Mussey  female  54.0      1   
497                  Shellard, Mr. Frederick William    male   NaN      0   
498  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0      1   
499                               Svensson, Mr. Olof    male  24.0      0   

     Parch     Ticket      Fare    Cabin Embarked  
250      0     362316    7.2500      NaN        S  
251      1     347054   10.4625       G6        S  
252      0     113514   26.5500      C87        S  
253      0  A/5. 3336   16.1000      NaN        S  
254      2     370129   20.2125      NaN        S  
..     ...        ...       ...      ...      ...  
495      0       2627   14.4583      NaN        C  
496      0      36947   78.2667      D20        C  
497      0  C.A. 6212   15.1000      NaN        S  
498      2     113781  151.5500  C22 C26        S  
499      0     350035    7.7958      NaN        S  

[250 rows x 12 columns]

With this basic syntax, you can lazily read in bits and pieces of a file at a time, without having to worry about being able to fit the entire file into RAM at once.  This is useful for doing some basic exploratory data analysis, and peek into a file, even if it's unreasonably large.

However, what if you wanted some way to go through the *entire* file, and make bulk changes to it?  

You can do that as well, by looping through the file stream. Here's a simple example, that prints off the total memory footprint for each chunk being read in:

In [7]:
# this basically reads: for every chunk in the file stream
for chunk in pd.read_csv('./data/titanic.csv', chunksize=200):
    # do this to each chunk
    print(chunk.memory_usage().sum())

19328
19332
19332
19332
8868


Notice what we're doing here.  `chunk` is the portion of the file that you're reading in and operating on.  `pd.read_csv(.....)` in this case is the iterator that you're looping over, not the entire file itself.

What's a little difficult to get used to is that `chunk` isn't some stand alone file that you play with once you've loaded it in.  It's just a portion of a loop that you pass some commands to.

To finish this section of the homework, you'll need to write functions that allow you to stream in data, and perform specific operations on your data set, without having it entirely loaded into memory, or being able to see it.

These functions should be written in a file called `chunking.py`, and each of the following two functions should be able to be called from an IDE to observe the results.

**Function 1**

**Name:** `probe_df`

**Arguments:** 
 - `file_path`, str; required,  Location of file to read in.
 - `chunksize`, int, required, default value is 1000.  Size of the chunk to use when streaming in the file.
 
**Returns:** a dictionary encoded in the following way: 
 - each key is the name of a column within your dataset
 - the value for each key is another dictionary with the following key/value pairs:
   - `null values`: number of null values for that column
   - `dtype`: data type for that column
   - `avg_val`: average value for that column ( if numeric, otherwise don't include )
   
**Note:** The `chunksize` argument can be used with a variety of file types, but you can just assume that you'll be reading a csv file, and nothing more.

**Function II**

**Name:** `write_df`

**Arguments:**

 - `file_path_read`   : str, required; location of file to read in.
 - `file_path_write`  : str, required; location of the file to write the new file out to
 - `chunksize`, int, required, default value is 1000.  Size of the chunk to use when streaming in the file.
 - `missing_vals`: dict, optional; accepts a dictionary as an argument with key/value pairs that list the column in the dataset(key) as well as the value to fill missing values with for that column(value).  The values in this dictionary will be used to fill missing values in the file at the location in `file_path_read`.
 
**Returns:**

This function will **not** return a value in the terminal.  What it **will** do is write a new file to the location specified in `file_path_write`.  

So, for example if you call `write_df('file/path/to/stream', 'file/path/to/write')`, a new file will appear in the location at `file/path/to/write` as the function is being called.

The big idea behind this file is that you'll be able to do data cleaning operations on a file you've never actually seen in your terminal before.

**Hint:** Pandas has a `to_csv()` method, with the option of appending lines to the end of it -- this is good to use when looping through the file stream.

**Note:** We will check to see that column headers are not added multiple times!

**To Test:** We will first call `probe_df` on the `taxi` dataset in its original location to get a dictionary with its missing values.  We will then use that value as a basis for the `missing_vals` argument in the `write_df` function, and use those to fill in its missing values.

If both of these functions work as intended, they will allow us to fill in the dataset's missing values without having looked at it, and you will receive full marks.